File: backend/app/services/resume_service.py
import base64
from io import BytesIO

from pypdf import PdfReader

from app.models.resume import ParseResumeRequest, ParseResumeResponse


def _infer_mime_type(file_name: str | None, mime_type: str | None) -> str:
    if mime_type and mime_type.strip():
        return mime_type

    lower_name = (file_name or "").lower()
    if lower_name.endswith(".pdf"):
        return "application/pdf"
    if lower_name.endswith(".txt"):
        return "text/plain"
    return "application/octet-stream"


async def parse_resume(payload: ParseResumeRequest) -> ParseResumeResponse:
    if not payload.dataBase64:
        raise ValueError("Missing resume file data.")

    try:
        file_bytes = base64.b64decode(payload.dataBase64, validate=True)
    except Exception as error:
        raise ValueError("Resume file data was not valid base64.") from error

    mime_type = _infer_mime_type(payload.fileName, payload.mimeType)
    if mime_type.startswith("text/"):
        try:
            text = file_bytes.decode("utf-8").strip()
        except UnicodeDecodeError as error:
            raise ValueError("Text resume was not valid UTF-8.") from error
        if not text:
            raise ValueError("Text resume was empty.")
        return ParseResumeResponse(text=text)

    if mime_type != "application/pdf":
        raise ValueError("Only PDF and text resumes are supported.")

    try:
        reader = PdfReader(BytesIO(file_bytes))
        pages = [page.extract_text() or "" for page in reader.pages]
    except Exception as error:
        raise ValueError("Could not read that PDF resume.") from error

    text = "\n\n".join(page.strip() for page in pages if page.strip()).strip()
    if not text:
        raise ValueError(
            "No text could be extracted from that PDF. Try a text-based PDF instead of a scanned image."
        )

    return ParseResumeResponse(text=text)

File: backend/src/services/resumeService.ts
import { generateGeminiText } from './gemini.js'

export type ParseResumeInput = {
  fileName?: string
  mimeType?: string
  dataBase64?: string
}

export type ParseResumeResult = {
  text: string
}

function inferMimeType(fileName?: string, mimeType?: string): string {
  if (mimeType?.trim()) return mimeType
  const lower = fileName?.toLowerCase() ?? ''
  if (lower.endsWith('.pdf')) return 'application/pdf'
  if (lower.endsWith('.txt')) return 'text/plain'
  return 'application/octet-stream'
}

export async function parseResume(input: ParseResumeInput): Promise<ParseResumeResult> {
  if (!input.dataBase64) {
    throw new Error('Missing resume file data.')
  }

  const mimeType = inferMimeType(input.fileName, input.mimeType)
  if (mimeType.startsWith('text/')) {
    const text = Buffer.from(input.dataBase64, 'base64').toString('utf8').trim()
    if (!text) throw new Error('Text resume was empty.')
    return { text }
  }

  if (mimeType !== 'application/pdf') {
    throw new Error('Only PDF and text resumes are supported.')
  }

  const text = await generateGeminiText({
    contents: [
      {
        role: 'user',
        parts: [
          {
            text:
              'Extract the complete resume text from this PDF. Return plain text only. Preserve section headers, bullet points, dates, company names, titles, and metrics. Do not summarize.',
          },
          {
            inline_data: {
              mime_type: 'application/pdf',
              data: input.dataBase64,
            },
          },
        ],
      },
    ],
    generationConfig: {
      temperature: 0,
    },
  })

  return { text }
}
